source/common/router/upstream_request.cc

```diff
@@ -432,6 +432,8 @@ void UpstreamRequest::acceptHeadersFromRouter(bool end_stream) {
   filter_manager_->requestHeadersInitialized();
   filter_manager_->streamInfo().setRequestHeaders(*parent_.downstreamHeaders());
   filter_manager_->decodeHeaders(*parent_.downstreamHeaders(), end_stream);
+  // Invoke the callback now that we've transmitted headers to the upstream.
+  this->parent_.onHeadersTransmittedToUpstream();
```

Comment:
I don't think we want to do this here.
If you have an upstream filter chain there may be pauses in the upstream filter chain, so while this will always be true if we've sent bytes upstream it will often be true if we haven't sent bytes upstream.

I think a more accurate place to check is in the UpstreamCodecFilter::decodeHeaders( which is conveniently where we set upstreamTiming().onFirstUpstreamTxByteSent which I think we can just use instead

Reply:
Thanks for the pointer! Agreed, decodeHeaders is a better place for this; will refactor